controllers/key_controller.go

```diff
@@ -0,0 +1,274 @@
+				err = r.GitClient.DeleteKey(ctx, key.Status.GitHubOrganization, key.Status.GitHubRepository, key.Status.GitHubKeyID)
+				if err != nil {
+					return ctrl.Result{RequeueAfter: requeueAfter}, err
+				}
+			}
+
+			r.updateKeyStatus(ctx, &key, v1alpha1.CreatingStatus)
+			log.Info("referenced secret does not exist, generating new key")
+			privateKey, privKeyErr := keygen.GenerateRSAPrivateKey(4096)
+			if privKeyErr != nil {
+				return ctrl.Result{RequeueAfter: requeueAfter}, privKeyErr
+			}
+			publicKey, publicKeyErr := keygen.GenerateRSAPublicKey(&privateKey.PublicKey)
+			if publicKeyErr != nil {
+				return ctrl.Result{RequeueAfter: requeueAfter}, publicKeyErr
+			}
+			privateKeyPem := keygen.EncodeRSAPrivateKeyToPEM(privateKey)
+
+			// Update the Key.status first
+			// When Secret creation fails, Reconcile will overwrite the old publicKey with a new one
+			// (If we created the Secret first but failed to update the status, it would orphan the Secret and error out the Key on next Reconcile.)
+			// note:
+			// 	This could cause a race condition if a client attempts to imperatively use the first value that is set as the PublicKey as it may be overwritten.
+			//  Clients should also check the Status value and potentially wait until the key is copied to GitHub.
+			// 	Reconcile will indefinitely update the PublicKey status with a new value while Secret creation is unauthorized.
+			if err := r.updateKeyStatusCreatingPublicKey(ctx, req.NamespacedName, string(publicKey)); err != nil {
+				return ctrl.Result{RequeueAfter: requeueAfter}, err
+			}
+
+			secret.SetName(secretRef.Name)
+			secret.SetNamespace(secretRef.Namespace)
+			secret.ObjectMeta.SetLabels(key.Spec.SecretTemplate.DeepCopy().Labels)
+			secret.ObjectMeta.SetAnnotations(key.Spec.SecretTemplate.DeepCopy().Annotations)
+			secret.Data = map[string][]byte{
+				"identity":     privateKeyPem,
+				"identity.pub": publicKey,
+			}
+			// TODO(stealthybox): v1.18 clients can make this Secret immutable -- should we?
+			// 	It might be useful but very weird to allow other clients to add keys to this Secret.
+
+			if secretRef.Namespace == req.Namespace {
+				// Owner refs only work in the same Namespace: https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/
+				// Queuing the owning Key based on the Secret changing or being deleted only works within a Namespace.
+				secret.SetOwnerReferences([]metav1.OwnerReference{asOwner(&key)})
+			}
+
+			if err := r.Client.Create(ctx, &secret); err != nil {
+				return ctrl.Result{RequeueAfter: requeueAfter}, err
+			}
+			log.Info("created new secret")
+
+			// Block until the cache updates, so that our Secret returns without failing the Get() on the next Reconcile
+			// If we don't do this, it breaks the publicKey matching logic between Key and the child Secret
+			// TODO(stealthybox): Can we remove this by using a non-cached client for Secrets CRUD?  Why does Create() not update the cache?
```

Comment:
Doesn't it do this because you aren't supposed to directly change the cache because then there is collisions so the idea is let everything be async. 

I wonder if you could get around this by exiting the reconcile and catching it after it's been created through the `Create()` of the owned resource? That probably doesn't work for cross Namespace setup but would work for anything within the same NS. You'd have to calculate the status upfront and operate based on where all the objects where in there loop…